viewing_party/party.py

```diff
@@ -1,23 +1,185 @@
 
-        
+# additional function to get the list of friends watched dictionaries
+def get_friends_watched_list(user_data):
+    friends = user_data["friends"]
+    friends_watched_list = []
+    for friend in friends:
+        for film in friend["watched"]:
+            friends_watched_list.append(film)
+    return friends_watched_list
+
+def get_unique_watched(user_data):
+    watched_movies = user_data["watched"]
+
+    # get the list of friends watched dictionaries
+    friends_watched_list = get_friends_watched_list(user_data)
+    
+    # get unique just by user watched movies
+    just_user_watched = []
+    for movie in watched_movies:
+        if movie not in friends_watched_list:
+            just_user_watched.append(movie)
+    return just_user_watched
+
+def get_friends_unique_watched(user_data):   
+    watched_movies = user_data["watched"]
+
+    # get the list of friends watched dictionaries
+    friends_watched_list = get_friends_watched_list(user_data)
+    
+    # get unique just by friends watched movies
+    just_friend_watched = []
+    for friend_movie in friends_watched_list:
+        if friend_movie not in watched_movies:
+            if friend_movie not in just_friend_watched:
+                just_friend_watched.append(friend_movie)
+    return just_friend_watched
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    # collect data: watched movies, list of subscriptions, friends watch list
+    watched_movies = user_data["watched"]
+    user_subscriptions= user_data["subscriptions"]
+    friends_watched_list = get_friends_watched_list(user_data)
+   
+    # make a recomendation list by subscription
+    recomendation_list = []
+    for friend_movie in friends_watched_list:
+        if friend_movie not in watched_movies:
+            if friend_movie["host"] in user_subscriptions:
+                recomendation_list.append(friend_movie)
+    return recomendation_list
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    # find user's most frequently watched genre, collect data: watched movies, friends watch list
+    watched_movies = user_data["watched"]
+    user_most_watched_genre = get_most_watched_genre(user_data)
+    friends_watched_list = get_friends_watched_list(user_data)
+    
+    # check for data not to be empty
+    recommended_movies = []
+    if len(watched_movies) == 0 or len(friends_watched_list) == 0:
+        return recommended_movies
+    else:
+    # make a recomendation list by genre
+        for friend_movie in friends_watched_list:
+            if friend_movie not in watched_movies:
+                if friend_movie["genre"] in user_most_watched_genre:
+                    recommended_movies.append(friend_movie)
+        return recommended_movies

```

Comment:
Checking whether the lists are empty is not necessary here; if a list is empty then a loop over that list will be skipped. The logic inside is great though!